File: packages/shared/response.ts
export const enum ResponseCode {
  Success = 656,
  FileNotFound,
  PermissionDenied,
  UnknownError,
}

export interface ResponseResult {
  code: ResponseCode;
  message: string;
}

export function createResponseResult(
  code: ResponseCode,
  message: string,
): ResponseResult {
  return {
    code,
    message,
  };
}

export function isResponseResult(result: unknown): result is ResponseResult {
  return (
    typeof result === 'object' &&
    result !== null &&
    Object.prototype.hasOwnProperty.call(result, 'code') &&
    Object.prototype.hasOwnProperty.call(result, 'message')
  );
}

File: crates/modules/shared/src/response.rs
use serde::{Deserialize, Serialize, Serializer};

#[derive(Deserialize, Clone, Copy, Debug)]
pub enum ResponseCode {
    Success = 656,
    FileNotFound,
    PermissionDenied,
    UnknownError,
}

impl Serialize for ResponseCode {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_i32(*self as i32)
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ResponseResult {
    pub code: ResponseCode,
    pub message: String,
}
